In [1]:
import pandas as pd
import psycopg2 as pg2
from sqlalchemy import create_engine

engine = create_engine('postgresql://testuser:testpass@localhost:5432/postgresql_analysis')

con = pg2.connect(host='localhost',
                  user='testuser',
                  password='testpass',
                  database='postgresql_analysis')
con.autocommit = True
cur = con.cursor()

In [2]:
def select(query):
    return pd.read_sql(query, con)

In [3]:
query = """
        SELECT * FROM reviews
        """

select(query)

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                 None   
1                 5                 None   
2                 5                 None   
3                 5                 None   
4                 5                 None   
...             ...                  ...   
99219             5                 None   
99220             5                 None   
99221             5                 None   
99222             4                 None   
99223             1                 None   

                                  review_comment_message review_creation_date  \
0                                                   None           2018-01-18   
1                                                   None           2018-03-10   
2                                                   None           2018-02-17   
3                  Recebi bem antes do prazo estipulado.           2017-04-21   
4      Parabéns lojas lannister adorei comprar pela I...           2018-03-01   
...                                                  ...                  ...   
99219                                               None           2018-07-07   
99220                                               None           2017-12-09   
99221  Excelente mochila, entrega super rápida. Super...           2018-03-22   
99222                                               None           2018-07-01   
99223  meu produto chegou e ja tenho que devolver, po...           2017-07-03   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

In [4]:
query = """
        SELECT * FROM order_items
        """

select(query)

order_id order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214             1   
1       00018f77f2f0320c557190d7a144bdd3             1   
2       000229ec398224ef6ca0657da4fc703e             1   
3       00024acbcdf0a6daa1e931b038114c75             1   
4       00042b26cf59d7ce69dfabb4e55b4fd9             1   
...                                  ...           ...   
112645  fffc94f6ce00a00581880bf54a75a037             1   
112646  fffcd46ef2263f404302a634eb57f7eb             1   
112647  fffce4705a9662cd70adb13d4a31832d             1   
112648  fffe18544ffabc95dfada21779c9644f             1   
112649  fffe41c64501cc87c801fd61db3f6244             1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

       shipping_limit_date   price  freight_value  
0      2017-09-19 09:45:35   58.90          13.29  
1      2017-05-03 11:05:13  239.90          19.93  
2      2018-01-18 14:48:30  199.00          17.87  
3      2018-08-15 10:10:18   12.99          12.79  
4      2017-02-13 13:57:51  199.90          18.14  
...                    ...     ...            ...  
112645 2018-05-02 04:11:01  299.99          43.41  
112646 2018-07-20 04:31:48  350.00          36.53  
112647 2017-10-30 17:14:25   99.90          16.95  
112648 2017-08-21 00:04:32   55.99           8.72  
112649 2018-06-12 17:10:13   43.00          12.79  

[112650 rows x 7 columns]

### Checking review data

Once the customer receives the product, or the estimated delivery date is due, the customer gets a satisfaction survey by email where he can give a note for the purchase experience and write down some comments.

In [5]:
query = """
        SELECT count(review_score) as review_score_null
        FROM reviews
        WHERE review_score = 0 or review_score is null
        """

select(query)

review_score_null
0                  0

Surprisingly, there's no 0 review score and every order has at least one review score. I want to know how many scores (1 to 5) there are in each order status group.

In [13]:
query = """
        SELECT order_status
            ,sum(case when review_score = 1 then 1 else 0 end) as score_1
            ,sum(case when review_score = 2 then 1 else 0 end) as score_2
            ,sum(case when review_score = 3 then 1 else 0 end) as score_3
            ,sum(case when review_score = 4 then 1 else 0 end) as score_4
            ,sum(case when review_score = 5 then 1 else 0 end) as score_55            
        FROM 
        (
            SELECT order_status, review_score
            FROM orders o
            JOIN reviews r on o.order_id = r.order_id
        ) a
        GROUP BY 1
        """

select(query)

order_status  score_1  score_2  score_3  score_4  score_55
0     approved        1        0        0        1         0
1     canceled      422       44       48       26        69
2      created        2        0        0        0         1
3    delivered     9406     2941     7961    18987     57066
4     invoiced      230       26       16       15        26
5   processing      256       18        9        6         7
6      shipped      644       79      110       87       123
7  unavailable      463       43       35       20        36

Are these review scores assigned before or after the deadline?

In [17]:
query = """
        SELECT order_status
            ,sum(case when 
                date_part('hour', age(review_answer_timestamp, order_estimated_delivery_date )) >= 0 
                then 1 else 0 end) as after_deadline
            ,sum(case when 
                date_part('hour', age(review_answer_timestamp, order_estimated_delivery_date)) < 0 
                then 1 else 0 end) as before_deadline
        FROM orders o
        JOIN reviews r on o.order_id = r.order_id
        GROUP BY 1
        """

select(query)

order_status  after_deadline  before_deadline
0     approved               2                0
1     canceled             517               92
2      created               3                0
3    delivered           20576            75785
4     invoiced             308                5
5   processing             293                3
6      shipped             990               53
7  unavailable             590                7

Now we see that customers can somehow review an order even when the deadline is not due.

And are these review scores assigned before or after the survey created?

In [79]:
query = """
        SELECT order_status
            ,sum(case when 
                date_part('hour', age(review_answer_timestamp, review_creation_date)) >= 0 
                then 1 else 0 end) as after_survey
            ,sum(case when 
                date_part('hour', age(review_answer_timestamp, review_creation_date)) < 0 
                then 1 else 0 end) as before_survey
        FROM orders o
        JOIN reviews r on o.order_id = r.order_id
        GROUP BY 1
        """

select(query)

order_status  after_survey  before_survey
0     approved             2              0
1     canceled           609              0
2      created             3              0
3    delivered         96361              0
4     invoiced           313              0
5   processing           296              0
6      shipped          1043              0
7  unavailable           597              0

Alright, now we're sure that these review scores are collected from satisfaction surveys. And these surveys are not sent out only when the order is delivered or after the deadline is due. Well, all customers really completed these surveys thoroughly.

### Average review score

In [93]:
query = """
        SELECT review_score
            ,count(distinct seller_id) as num_seller_by_review
        FROM reviews r
        JOIN order_items oi on r.order_id = oi.order_id
        GROUP BY 1
        ORDER BY 1
        """
select(query)

review_score  num_seller_by_review
0             1                  1774
1             2                   929
2             3                  1420
3             4                  1980
4             5                  2700

In [95]:
query = """
        SELECT seller_id, avg(review_score) as avg_score
        FROM reviews r
        JOIN order_items oi on r.order_id = oi.order_id
        GROUP BY seller_id
        """
select(query)

seller_id  avg_score
0     a61cc04793308395a840807104365121   5.000000
1     6d2f2e3b539480db1e0842b3a4e32e6e   4.333333
2     c53bcd3be457a342a97e39e5a9f0be22   3.200000
3     9591fc341b1bfb7ef561e2968ec6e011   1.000000
4     f9ec7093df3a7b346b7bcf7864069ca3   4.777778
...                                ...        ...
3085  b2eecf5ea250510da76590ca79d60e5d   5.000000
3086  eae9af4811c294c56795d70e715b7337   5.000000
3087  1c83418358f249c9d6affa52d579db0e   5.000000
3088  845e8036c3f7bf62099352f5ed910240   3.833333
3089  01cf7e3d21494c41fb86034f2e714fa1   4.212121

[3090 rows x 2 columns]

In [96]:
query = """
        SELECT b.num_sellers
            ,count(a.seller_id) as num_qualified_sellers
            ,round(count(a.seller_id) * 1.0 / b.num_sellers * 100, 2) as percentage
        FROM 
        (
            SELECT seller_id, avg(review_score) as avg_score
            FROM reviews r
            JOIN order_items oi on r.order_id = oi.order_id
            GROUP BY seller_id
            HAVING avg(review_score) >= 4
        ) a
        CROSS JOIN 
        (
            SELECT count(distinct seller_id) as num_sellers
            FROM order_items
        ) b
        GROUP BY 1
        """

select(query)

num_sellers  num_qualified_sellers  percentage
0         3095                   1992       64.36

64.36% sellers have average review score of 4 or higher. Most customers was satisfied with their purchase, giving a rating of 4 or higher.